### PR TITLE
Port to GTK4

### PR DESCRIPTION
**Pull request by treppenwitz03**
Goes with https://github.com/elementary/switchboard/pull/231

### CODE REVIEW

**Comment by jeremypw** [comment]
There is a difference in alignment of buttons in the Add Print dialog otherwise this seems to match the existing UI well.

 **NEW**
Screenshot from 2022-07-11 11 03 06

**OLD**
Screenshot from 2022-07-11 11 02 36

**Comment by jeremypw** [comment]
Do references to "gtk30" need changing?

**Comment by ZenitsuDev** [comment]
sorry for the late response @jeremypw . I will try to update this later. Thank you

**Comment by treppenwitz03** [comment]
> There is a difference in alignment of buttons in the Add Print dialog otherwise this seems to match the existing UI well.
> 
> **NEW** Screenshot from 2022-07-11 11 03 06
> 
> **OLD** Screenshot from 2022-07-11 11 02 36

This should be fixed now.

**Comment by treppenwitz03** [comment]
> There is also a difference in wrapping of the clear all confirmation dialog (at least in English):
> **NEW**
> Screenshot from 2022-07-11 11 07 48
> **OLD**
> Screenshot from 2022-07-11 11 07 30

ClearQueueDialog subclasses directly from Granite.MessageDialog so it may be the cause of the wrapping issue.

**Comment by treppenwitz03** [comment]
> This ink levels seem to be drawn in accent color regardless of the actual ink color. I faked a printer with only blue (#0000AA) ink:
> 
> **NEW** Screenshot from 2022-07-11 11 30 33 **OLD** Screenshot from 2022-07-11 11 30 38

Pardon me about this one, but maybe someone might be able to find the fault in here? I do not a printer right now to be able to test changes I will make. However, looking at the code, the fault I think may lie in the STYLE_CLASS constant in src/Widgets/InkLevel.vala as there might be a change in css node of Gtk.LevelBar.

File: src/Views/SetupView.vala [hunk 1]
```diff
@@ -32,7 +32,7 @@ public class Printers.OptionsPage : Gtk.Grid {
         var pages_per_sheet = new Gee.TreeSet<int> ();
         var default_page = printer.get_pages_per_sheet (pages_per_sheet);
 
-        var pages_modebutton = new Granite.Widgets.ModeButton ();
+        var pages_modebutton = new Printers.ModeButton ();
```
Comment (danirabbit) [inline src/Views/SetupView.vala:35]: This should be radiobuttons instead of a custom widget

**Comment by danirabbit** [comment]
I think this is ready for another re-review! @jeremypw if you're still interested, let me know if you think this needs to be broken up more to be easier to review.

**Review by jeremypw — CHANGES_REQUESTED** [review]
I have not yet tested all functionality but notice that the "Add Printer" button remains insensitive after choosing a printer (in master it becomes sensitive under the same conditions)

**Comment by danirabbit** [comment]
@jeremypw figured out the source of the problem there and ended up rewriting this a bit to use validated entries and give more feedback etc https://github.com/elementary/switchboard-plug-printers/pull/197

**Comment by danirabbit** [comment]
> Another issue I noticed is that pressing the "Clear All" button only cleared the last job, not all of them.

Fixed! headers and placeholders are considered children here so switched these listbox iterators to use `get_row_at_index` which only includes rows and not headers or placeholders

**Comment by jeremypw** [comment]
The inklevel indicators do not expand vertically, unlike master so they appear unnecessarily squat (the inks are faked as I do not have a color printer). 

Screenshot from 2023-11-01 19 23 09
Screenshot from 2023-11-01 19 23 15

**Comment by jeremypw** [comment]
I noticed then when printing a test page to a CUPS_BRF PDF printer, the settings for Color/GreyScale and for Portrait/Landscape seemed to have no effect although reversing did.  However the same thing happens in master so not a blocker here.

**Comment by danirabbit** [comment]
@jeremypw I can also confirm that those settings don't seem to do anything for a PDF printer, but like you said the problem also exists in the current master and that code wasn't touched in this branch

It looks like the other issue with the ink levelbars was fixed though!

**Review by jeremypw — CHANGES_REQUESTED** [review]
Nearly everything works except removing a printer does not actually remove it until the plug is closed (regression from master which removes the printer and refreshes the list straightaway.

**Review by jeremypw — APPROVED** [review]

**Pull request merged**